Resuming an interrupted `git webkit pr` produces a commit message without links

briannafan opened the pull request:
#### c2b3cb0e470b3a305b95c7c82557390653433235
<pre>
Resuming an interrupted `git webkit pr` produces a commit message without links
<a href="https://rdar.apple.com/125444927">rdar://125444927</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=271279">https://bugs.webkit.org/show_bug.cgi?id=271279</a>

Reviewed by Jonathan Bedard.

When creating a commit in program.commit or program.pull_request, we shouldn&apos;t
set env variables if the bug title and urls are null.

If these are empty, we check for the variables stored in the branch config.

* Tools/Scripts/hooks/prepare-commit-msg:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/commit.py:
(Commit.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_commit):

Canonical link: <a href="https://commits.webkit.org/289804@main">https://commits.webkit.org/289804@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e40451b1c10506944cc948e3e2f563025878224b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87448 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92310 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38187 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89499 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7343 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15126 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67545 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25284 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90450 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79126 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47886 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/86944 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5379 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33520 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37303 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75797 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94195 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14613 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10717 "Found 2 new test failures: fast/css/viewport-height-border.html media/video-replaces-poster.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76368 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/87033 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14867 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74981 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75585 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19964 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18383 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7551 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13739 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14631 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14374 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17818 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16157 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -44,7 +44,8 @@ def set_env_variables_from_branch_config():
         variable, mode = data
         if variable in os.environ:
-            continue
+            if os.environ[variable]:
```
Surely they're not None, per the commit message, but just the empty string? How come they end up still being set, just to the empty string?

Like, I don't understand how this fixes the bug.

briannafan replied:
I think that's my bad for writing null when I meant empty. 
In `commits.py`:
```
env['COMMIT_MESSAGE_TITLE'] = issue.title if issue else ''
env['COMMIT_MESSAGE_BUG'] = '\n'.join(cls.bug_urls(issue))
```
This fixes the bug because we'll check the branch config if either these env variables don't exist or if they are the empty string.

briannafan replied:
Previously, we'd skip checking the branch config when the variables are empty, thus populating the message with empty bug urls.

gsnedders replied:
Ah, but the `git-webkit pr` case is defining them in `webkitscmpy/program/pull_request.py`, not `program/commit.py` (honestly this code is rather duplicated and we ought to de-duplicate this):

https://github.com/WebKit/WebKit/blob/75189e2dbaae35bea7231e1b00fb9ee31f60412d/Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py#L171-L172

And we unconditionally set these, even though we only *also* set `args._title` and `args._bug_urls` when we've just created a branch, thus these end up equal to the empty string.

I think my first comment is we shouldn't be setting these to the empty string — that's just not useful, we want `prepare-commit-msg` to use its own logic to try and find them (and these will be set in `git config branch.*.title` etc. from when we created the branch on the first invocation); I don't know if it makes sense to distinguish between empty and unset, especially in light of a request like https://bugs.webkit.org/show_bug.cgi?id=282648 to make these more user-exposed.

So I think I'd rather we make this PR:

1. Only set `COMMIT_MESSAGE_TITLE `/`COMMIT_MESSAGE_BUG` when non-empty (everywhere they're set!).
2. Change this `if variable in os.environ: if os.environ[variable]: continue` to `if variable in os.environ and os.environ[variable]: continue`

briannafan replied:
I filed a bug about the duplication! I agree with your assessment, I briefly thought about it but took the easy way out :P

briannafan committed:
Resuming an interrupted `git webkit pr` produces a commit message without links
rdar://125444927
https://bugs.webkit.org/show_bug.cgi?id=271279

Reviewed by Jonathan Bedard.

When creating a commit in program.commit or program.pull_request, we shouldn't
set env variables if the bug title and urls are null.

If these are empty, we check for the variables stored in the branch config.

* Tools/Scripts/hooks/prepare-commit-msg:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/commit.py:
(Commit.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_commit):

Canonical link: https://commits.webkit.org/289804@main
